### ISSUE TITLE
Monkey Class type

### ISSUE THREAD

**Issue by WeyrSDev** [bmx-ng/bmx-ng#63]
Could they not be brought into Bmax? i tried the full version for first time and being able to use private/protected or is that to big a change?

Brucey you dont happen to have any demos using vulkun api? i cant find much on it for blitz.

there a forum for NG?

**Comment by HurryStarfish** [comment]
Can you clarify what you mean by "monkey class"?
You can use `Private`/`Protected` inside classes in BlitzMax-NG. They will apply to all members following them and restrict their visibility/accessibility from other files. Private/protected members will still be accessible in the same file the class is declared in, though.

**Comment by llYuell** [comment]
It would be good to create an exclusive discord chat channel for NG.

**Comment by WeyrSDev** [comment]
i mean the syntax and function of monkeys-x Class usage 
i.e. blitzmax
Type MyType
       Field SomeSuff:String

       Function Create:MyType()
          Return New MyType
       End Function
End Type

In Monkey
Class MyClass
       Method New(); End Method
Protected
       Field gMyType:MyType '// using the type as a struct
End Class

I if i had the know how i would just use Types as Structs and use Class for objects every one knows what a class represents and i think it just flows better syntax wise.

TLDR
Keep Blitz max Types but use them like you would a struct and add class / End class functionality.

I basically want Monky and Blitz max Syntax to merge but using Blitz max as the main system.

we could have :bool types then also

I know about Monkey2 but its not the same and the syntax feels so much like c++ i might as well use that. Blitz Max is near perfect  but adding the good bits from monkey-x would make it better than Monkey2.

I have more faith in brucey to get things done than Mark he just keeps making new when Blitz Max was perfect in difficulty and syntax style  just missing a few things

**Comment by woollybah** [comment]
Hallo :-)

I don't think we'll be changing Type to Class any time soon - it would break everything anyone has ever written in BlitzMax.

With Blitzmax NG you should be able to accomplish most things that you can in Monkey, albeit that Type=Class.
We have a `Struct` type for structs., method overloading (including `New`) and operator overloading. There's also proper Enums coming soon.

We are always happy to consider feature requests, but I doubt a major change like that would be forthcoming, I'm afraid. The only thing I can imagine might be to have `Class` as an alias for `Type`...

**Comment by WeyrSDev** [comment]
ok cool any usage code for structs? also does that mean we can bring in c++ structs now? i made a math lib but its all structs with inline functions but i recall BlitzMax had limitations on this.

would love to see any example code you have testing new features like sdl/bgfx vulkun and any tests on importing c++ class/functions etc.

There needs to be a place to chat as i seem to be asking old questions but its hard to find info on changes upcoming and when its worth doing a full update

**Comment by llYuell** [comment]
Bearing in mind that BlitzMax NG has great capacity and power and has taught me quite a few things for object-oriented programming, it would be interesting to have an alias called Class, I think many people would see it more accessible and eye-catching than types.

**Comment by HurryStarfish** [comment]
I don't think a `Class` alias for `Type` would make things better - having two different keywords for the same thing would create more confusion. Having `Type` as the keyword for classes might be a bit unfortunate, but it must stay for backwards compatibility reasons. Adding `Class` on top of it would not get rid of it, it would only create a split between people and code that use `Type` and those that use `Class`, and I feel that would be much more confusing.

As for structs:
If you know C++, the easiest way to explain it is that the difference between a class ("`Type`") and a `Struct` in BlitzMax-NG is similar to the difference between C++ objects created with `new` and without `new`, respectively.
To be more precise: Structs are a sort of more "lightweight" alternative to classes. Unlike class instances/objects, struct instances are not managed by the garbage collector; and when you store them in variables, you're not storing a reference to the instance but the instance itself (similarly to primitive types like `Int`). This means they can be more efficient than classes, but in exchange, they have certain limitations (most importantly, they do not support subtyping/inheritance). They're mostly suited for grouping a bunch of values together in small, simple (ideally immutable) types, such as points, vectors, colors etc.
The C# programming language has the same distinction between structs and classes as BlitzMax-NG. So if you want, you can look for info on the topic in C#, too.

**Comment by llYuell** [comment]
@HurryStarfish 

But for some reason I perceive that an alias of Type, is Struct, and the most reasonable thing is that it is wrong, however that we have a type an alias called Struct.

**Comment by HurryStarfish** [comment]
`Struct` is not an alias of `Type`, see my previous comment for an explanation.

**Comment by llYuell** [comment]
I repeat, in my ignorance I see this as well.  I can even put functions and methods into a structure.

```

Struct T

	Field s% = 100

End Struct 


Local s:T = New T

s.s = 200

Print s.s
```
